contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,259 @@
+
+If you are concerned about appearances, you can always opt to [hide the sausage making](#hide-the-sausage-making) to clean up your commit history.
+
+If you have made several overlapping changes on your working directory (e.g. forgot to commit, etc), you can always [perform a patch-wise stage using `git add -p`](https://stackoverflow.com/questions/1085162/commit-only-part-of-a-file-in-git).
+
+Read more:
+- [programming practices - When to commit code?](https://softwareengineering.stackexchange.com/questions/83837/when-to-commit-code)
+- [Do commit early and often](https://sethrobertson.github.io/GitBestPractices/#commit)
+- [When is a version control commit too large?](https://softwareengineering.stackexchange.com/questions/10793/when-is-a-version-control-commit-too-large)
+- [Commit only part of a file in Git](https://stackoverflow.com/questions/1085162/commit-only-part-of-a-file-in-git) _this is useful to use as a cheatsheet during interactive staging_
+
+---
+
+## Hide the Sausage Making
+_Sausage making_ refers to the process by which code is incrementally worked on, where a series of commits (like links in a sausage) make up a branch. It is often desirable to _hide the sausage making_ where the commit history is cleaned up so that it looks neater and is easier to follow.
+
+When working on a `feature`/`fix` branch, you may
+- [commit non-functioning code](#always-commit-functional-code), or
+- may need [multiple commits to implement one logical change in code](#one-logical-change-to-one-commit), or
+- cherry-pick a lot of commits from other branches
+
+This may clutter your history with low-level details or make it convoluted to follow for a maintainer or reviewer. Like sausage, you may enjoy eating it but not the process of making it.
+
+### How to Hide the Sausage
+Hiding the sausage is typically achieved by either/both:
+- performing an interactive rebase, i.e. `git rebase -i`
+- patch-wise reset and stage, i.e. `git reset -p` and `git add -p`
+
+Ensure that you do this _before_
+- pushing to a remote repository [to respect the published history](#respect-published-history) and
+- performing any merges from other branches.
+
+Read more:
+- [On Sausage Making](https://sethrobertson.github.io/GitBestPractices/#sausage)
+- [Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) (great reference for how to actually do this, under _Changing Multiple Commit Messages_)
+
+---
+
+## Respect Published History
+Always avoid rewriting the public, published history unless you have a very good reason, you are very sure of what you are doing, or you will not affect anyone else.
```

Comment:
Maybe you could say `Always avoid rewriting the public history` or `Always avoid rewriting the published history` instead? Putting both terms in there seems redundant.